### PR TITLE
Support copying the overlapping region from one buffer to another.

### PR DESCRIPTION
**Pull request by mcourteaux**
TLDR: This PR makes it possible to copy from a smaller buffer to a bigger buffer as well (i.e. blit a buffer onto another buffer). Use case: workaround for #8235: realize pipeline to a buffer that acts as a crop, and then copy the result back.

---

This PR makes it possible to copy from any buffer to any other buffer with the same number of dimensions. This is in contrast to the original functionality where it tries to copy from src to the entire number of elements in the dst buffer (even if src is smaller).

 - Added a `dst_begin` in `device_copy`.
 - @derek-gerstmann I modified a few things in the Vulkan backend cleaning up those multidimensional copies (the tests passed on my machine, but please have a look at what I did in that backend if you can). I also brought it more in line with the other backends.
 - I brought the Metal backend also more in line with the others regarding how we deal with the `src_begin` and `dst_begin` values.
 - I added tests to test this behavior.
 - I validated it myself for my own pipelines, as a application-side workaround to deal with #8235 where I make new a new buffer that acts as a crop (but is not), and copy the result back to the original buffer after the pipeline is done.

Irrelevant: my clang-format for some reason wanted to move two things in Halide::Runtime::Buffer, which is compatible with the one on the automatic test here. It seems to be the difference between clang-format-17 (github test) and clang-format-18 (my machine).

### CODE REVIEW

**Comment by mcourteaux** [comment]
One of the build bots died? Can we restart one build specifically manually? Or will everything have to rerun?

**Comment by steven-johnson** [comment]
> One of the build bots died? Can we restart one build specifically manually? Or will everything have to rerun?

Sorry, I had to restart the buildmaster. You may need to rerun everything :-/

**Comment by mcourteaux** [comment]
I ran into a crash on my application. Will try to figure out what that is before I reopen for review.

File: src/runtime/HalideBuffer.h
Comment (zvookin) [inline src/runtime/HalideBuffer.h:1]: This is just to satisfy the formatter right?

File: src/runtime/cuda.cpp [hunk 1]
```diff
@@ -1010,7 +1010,9 @@ WEAK int halide_cuda_buffer_copy(void *user_context, struct halide_buffer_t *src
             }
         }
 
-        auto result = cuda_do_multidimensional_copy(user_context, c, c.src + c.src_begin, c.dst, dst->dimensions, from_host, to_host, stream);
+        auto result = cuda_do_multidimensional_copy(
+            user_context, c, c.src + c.src_begin, c.dst + c.dst_begin,
+            dst->dimensions, from_host, to_host, stream);
```
Comment (zvookin) [inline src/runtime/cuda.cpp:1015]: Adds offset for destination as well as source.

File: src/runtime/HalideBuffer.h
Comment (mcourteaux) [reply src/runtime/HalideBuffer.h:1]: I actually think the formatter doesn't care, but nvim puts it automatically when you save a text file. In Unix systems, a "line" is _defined_ as characters ended by a newline. The that last newline should sort of be there by definition of a line, which is what nvim does.

File: src/runtime/cuda.cpp (hunk 1 above)
Comment (mcourteaux) [reply src/runtime/cuda.cpp:1015]: Yes?

**Comment by mcourteaux** [comment]
Update, I tested this again, and I got this:

Error: OpenCL error: CL_MEM_OBJECT_ALLOCATION_FAILURE buffer copy failed

This seems the correct error for this case. Will reopen.

File: src/runtime/cuda.cpp (hunk 1 above)
Comment (zvookin) [reply src/runtime/cuda.cpp:1015]: There are formatting changes so I commented to make the logic change more obvious.

**Comment by mcourteaux** [comment]
All tests passing. Can we land this?

**Review by abadams — APPROVED** [review]
